### PR TITLE
Fix warnings found by GCC

### PR DESCRIPTION
**Pull request by jfbastien**
My previous patch addressed all LLVM warnings, this one addresses all the GCC ones as well (mostly signed / unsigned mix).

The patch also turns on -Wall -Werror.

### CODE REVIEW

File: src/asm2wasm.h [hunk 1]
```diff
@@ -358,7 +358,7 @@ class Asm2WasmBuilder {
     abort(); // avoid warning
   }
 
-  unsigned bytesToShift(unsigned bytes) {
+  int64_t bytesToShift(unsigned bytes) {
```
Comment (kripken) [inline src/asm2wasm.h]: 32 bits is enough, this is an asm.js pointer

File: src/compiler-support.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+#ifndef __has_feature
+# define __has_feature(x) 0
+#endif
+
+#ifndef __has_builtin
+# define __has_builtin(x) 0
+#endif
+
+// If control flow reaches the point of the WASM_UNREACHABLE(), the program is
```
Comment (kripken) [inline src/compiler-support.h]: oh nice, now I know the proper way to do this stuff :)

File: src/asm2wasm.h (hunk 1 above)
Comment (jfbastien) [inline src/asm2wasm.h]: Done.

File: src/s2wasm.h [hunk 3]
```diff
@@ -886,7 +886,9 @@ class S2WasmBuilder {
         zero = false;
       } else if (match(".zero")) {
         int32_t size = getInt();
-        for (size_t i = 0; i < size; i++) {
+        if (size <= 0)
```
Comment (kripken) [inline src/s2wasm.h]: curly braces, or same line (or does clang-format not accept either?)

File: src/compiler-support.h (hunk 2 above)
Comment (jfbastien) [inline src/compiler-support.h]: This is pretty new, older GCC doesn't haz feature and haz builtin... It falls back to abort even though it does have __builtin_unreachable. I could feature-detect GCC, but meh.

File: src/s2wasm.h (hunk 3 above)
Comment (jfbastien) [inline src/s2wasm.h]: clang-format doesn't add braces, I just need to get used to it :-)

**Pull request merged**